### ISSUE TITLE
Concat consecutive messages by same user?

### ISSUE THREAD

**Issue by etjossem** [hexchat/hexchat#1408]
When a message longer than the limit is typed, many clients will split them up and deliver as follows:

Etjossem: This is a long message this is a long message this is a long message this is a long message this is a long message this is a long message this is a long message this is a long message this is a long message this is a long message this is a long me
Etjossem: ssage, now it's done.

Would we want to concat such messages when delivered in short succession?

**Comment by Poorchop** [comment]
I think this has to do with the IRC protocol. I believe that the character length is 512 characters including the CRLF line ending, so the message being split locally is staying true to how it would be displayed to other clients. For long incoming messages, I'm not sure if there's any fail-proof way to concatenate them, although I guess the timestamps would be the same.

**Comment by TingPing** [comment]
The only way to handle it would just be *always* combining repeated messages from a user. Since for various reasons we can't actually know if two lines was intended to be sent as one.

**Comment by etjossem** [comment]
The way I imagined it: if the user is the same, compare incoming messages' timestamps and see if they're equal (margin of error is up to us), concatenate the messages if they are.

This would allow us to comply with the IRC protocol (512 characters max) but appear to allow unlimited message length.

I wouldn't concat *all* consecutive messages from the same user, since people often use multiple messages to denote paragraphs.

**Comment by TingPing** [comment]
> if the user is the same, compare the two consecutive incoming messages' timestamps and see if they're equal (margin of error is up to us)

We are talking about a networked application you cannot reason about timestamps of received messages and expect it to be reliable.

> Same goes for outgoing messages longer than 512 characters - split them on the way out as usual, but locally display them as though they were the same message.

That is technically doable. I am personally not a fan of any of these things as they only hide information from the user.

> I wouldn't concat all consecutive messages from the same user, since people often use multiple messages to denote paragraphs

I didn't mean directly concat, but show them as paragraphs without duplicate nick entries basically.

**Comment by etjossem** [comment]
> We are talking about a networked application you cannot reason about timestamps of received messages and expect it to be reliable.

Yes, this will only work if we have a timestamp for when it was *sent*, not just when it was received. I don't know enough about the protocol to say if that's available.

> That is technically doable. I am personally not a fan of any of these things as they only hide information from the user.

I wouldn't advise forcing this on all users or even turning it on by default. Make it a config option. Some people like the "friendly" style of longer paragraphs, others want it to more truly reflect the protocol.

> I didn't mean directly concat, but show them as paragraphs without duplicate nick entries basically.

I support this; it's the most straightforward way to make this improvement.

**Comment by Poorchop** [comment]
>Yes, doing this for inbound messages will only work if we have a timestamp for when they were sent, not just when they were received.

Timestamp information isn't provided at all as per the protocol. I'm pretty sure that's something that is always determined locally when new data is received or printed to the text area.

For inbound messages, I kind of thought the best way would be to look at string length and if it's >= 512, only then should the timestamps be compared. There is still a very remote chance that someone could meet the defined criteria for concatenating two messages that should actually be kept separate.

I also don't like the idea of showing multiple consecutive messages as paragraphs without the duplicate nick entries. Having a config for concatenating long outbound messages isn't a bad idea though.

**Comment by etjossem** [comment]
>=512 sounds like a good way to do it. Plus similar timestamps. Have it as an option, not default.
